File: Java/src/com/sonicscholar/simpleBoardGame/treasureHunt/TreasureHuntAIPlayer.java
package com.sonicscholar.simpleBoardGame.treasureHunt;

import com.sonicscholar.simpleBoardGame.AbstractPlayer;
import com.sonicscholar.simpleBoardGame.Board;
import com.sonicscholar.simpleBoardGame.DefaultGameAction;
import com.sonicscholar.simpleBoardGame.GameAction;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class TreasureHuntAIPlayer extends AbstractPlayer {
    private Board _board;

    public TreasureHuntAIPlayer(String name, char marker, Board board) {
        super(name, marker);
        _board = board;
    }

    @Override
    public boolean isComputer() {
        return true;
    }

    /**
     * Tallies up the empty spaces, and selects from them,
     * a random coordinate
     * @return GameAction for the AI Player
     */
    @Override
    public GameAction requestAction() {

        List<Coordinate> emptySpaces = new ArrayList<>();

        //tallying the empty spaces
        for (int row = 0; row < _board.getHeight(); row++){
            for (int col = 0; col < _board.getWidth(); col++) {
                boolean isEmpty = _board.isEmpty(row, col);
                if(isEmpty) {
                    Coordinate coordinate = new Coordinate(row, col);
                    emptySpaces.add(coordinate);
                }
            }
        }

        //pick an empty at random
        int numEmptySpaces = emptySpaces.size();
        Random random = new Random();
        int randomIndex = random.nextInt(numEmptySpaces);
        Coordinate desiredCoordinate = emptySpaces.get(randomIndex);
        int row = desiredCoordinate.getRow();
        int col = desiredCoordinate.getColumn();
        return new DefaultGameAction(row, col, _marker);
    }
}

File: C_Sharp/SonicScholar.SimpleBoardGame.Framework/TreasureGame/TreasureHuntAIPlayer.cs
using System;
using System.Collections.Generic;
using SonicScholar.SimpleBoardGame.Framework;

namespace TreasureGame
{
    public class TreasureHuntAIPlayer : PlayerBase
    {
        
        public TreasureHuntAIPlayer(string name, char marker, IBoard board) : base(name, marker)
        {
            Board = board;
        }

        internal IBoard Board { get; set; }

        public override bool IsComputer => true;
        
        public override IGameAction RequestAction()
        {
            List<Coordinate> emptySpaces = new List<Coordinate>();
            //tally up empty spaces
            for (int row = 0; row < Board.Height; row++)
            {
                for (int col = 0; col < Board.Width; col++)
                {
                    if (Board.IsEmpty(row, col))
                    {
                        Coordinate emptySpace = new Coordinate(row, col);
                        emptySpaces.Add(emptySpace);
                    }
                }
            }

            //randomly select space
            Random random = new Random();
            int randomIndex = random.Next(emptySpaces.Count);
            Coordinate desiredCoordinate = emptySpaces[randomIndex];

            return new GameAction(desiredCoordinate.Row, desiredCoordinate.Column, Marker);
        }
    }
}